File: backend/models/delivery.py
from mongoengine import Document, StringField, FloatField, IntField, DateTimeField, ReferenceField
import datetime
from .user import User
from .order import Order

class DeliveryBid(Document):
    deliveryPerson = ReferenceField(User, required=True)
    order = ReferenceField(Order, required=True)

    bid_amount = FloatField(required=True) # proposed bid amount
    status = StringField(
        choices = ["Pending", "Accepted", "Rejected"],
        default="Pending"
    )

    created_at = DateTimeField(default=datetime.datetime.utcnow)

    meta = {
        'collection': 'delivery_bids',
        'indexes': [
            "order",
            "deliveryPerson",
            "status"]
    }

class Delivery(Document):
    order = ReferenceField(Order, required=True)
    deliveryPerson = ReferenceField(User, required=True)

    bidAmount = FloatField(required=True)

    status = StringField(
        choices=["Assigned", "Out_For_Delivery", "Delivered", "Delivery_Failed"],
        default="Assigned"
    )
    note = StringField(default="")

    created_at = DateTimeField(default=datetime.datetime.utcnow)
    updated_at = DateTimeField(default=datetime.datetime.utcnow)

    meta = {
        "collection": "deliveries",
        "indexes": [
            "order",
            "deliveryPerson",
            "status"]
        }

    def set_status(self, new_status, note=None):
        self.status = new_status
        self.updated_at = datetime.datetime.utcnow()
        if note:
            self.note = note
        self.save()

    def attach_note(self, text):
        self.note = text
        self.save()

File: backend/models/Delivery.js
const mongoose = require('mongoose');

const deliveryBidSchema = new mongoose.Schema({
  deliveryPerson: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true
  },
  order: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Order',
    required: true
  },
  bid_amount: {
    type: Number,
    required: true
  },
  status: {
    type: String,
    enum: ["Pending", "Accepted", "Rejected"],
    default: "Pending"
  },
  created_at: {
    type: Date,
    default: Date.now
  }
}, {
  collection: 'delivery_bids'
});

deliveryBidSchema.index({ order: 1 });
deliveryBidSchema.index({ deliveryPerson: 1 });
deliveryBidSchema.index({ status: 1 });

const deliverySchema = new mongoose.Schema({
  order: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Order',
    required: true
  },
  deliveryPerson: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true
  },
  bidAmount: {
    type: Number,
    required: true
  },
  status: {
    type: String,
    enum: ["Assigned", "Out_For_Delivery", "Delivered", "Delivery_Failed"],
    default: "Assigned"
  },
  note: {
    type: String,
    default: ""
  },
  created_at: {
    type: Date,
    default: Date.now
  },
  updated_at: {
    type: Date,
    default: Date.now
  }
}, {
  collection: 'deliveries'
});

deliverySchema.index({ order: 1 });
deliverySchema.index({ deliveryPerson: 1 });
deliverySchema.index({ status: 1 });

deliverySchema.methods.set_status = function(new_status, note) {
  this.status = new_status;
  this.updated_at = new Date();
  if (note) {
    this.note = note;
  }
  return this.save();
};

deliverySchema.methods.attach_note = function(text) {
  this.note = text;
  return this.save();
};

module.exports = {
  DeliveryBid: mongoose.model('DeliveryBid', deliveryBidSchema),
  Delivery: mongoose.model('Delivery', deliverySchema)
};
